_posts/2020-04-03-rendering-in-UE4.md

```diff
@@ -0,0 +1,32 @@
+---
+author: "alepaoletti"
+layout: post
+did: "blog10"
+title: "Rendering in Unreal Engine 4"
+slug: "RenderingInUE4"
+date: 2020-04-06 08:00:00
+categories: rendering
+img: /banners/GPU_profiler.png
+banner: /banners/GPU_profiler.png
+tags: graphics
+description: "A really quick overview of the rendering pipeline in Unreal Engine 4."
+---
+
+Unreal Engine 4 has a lot of code for doing rendering: it exists in its own module - the *Renderer* module - which is compiled to a dll to allow faster iteration. The *Renderer* module depends on the Engine module because it has many callbacks into Engine. However, when the Engine needs to call some code in the Renderer, this happens through the *IRendererModule* interface.
+
+The *RHI* module is the other key module for graphics programming as it is the interface for rendering APIs. Here is where the magic happens to allow the abstract rendering code to work on different platforms using different APIs.
+
+The renderer code runs in a separate thread, the *Rendering Thread*. It operates in parallel with the game thread and it's usually one or two frames behind it. It serves to enqueue platform-agnostic render commands into the renderer's command list through the `ENQUEUE_RENDER_COMMAND` macro. To make sure your code is called by the right thread, you can add checks such as `check(IsInGameThread())` or `check(IsInRenderingThread())` for improving code stability.
+
+Finally, a new thread, the *RHI Thread*, executes these commands via the appropriate graphics API on the backend.
+
+It's a complex topic, especially if you want to make modifications to customize it for your own purposes.
+
+Here is the official UE4 documentation: [UE4 Graphics Programming](https://docs.unrealengine.com/en-US/Programming/Rendering/index.html).
+
+### Why am I writing this
```

Comment:
This is a nice conclusion section. But what about adding some more content in this article, before this section?
I mean, you simply described the high-level architecture of the pipeline - do you have some other quick tip you can add to this article (maybe with some code snippet) to make it more useful for the reader (and help the search engines too)?

Reply:
I definitely do. I wrote this just as an introduction to something that I would like to keep doing.

Reply:
Do you intend to do it in this article, or in other articles?

Reply:
I'm about to change a little the content of this post. It will take a while, but I'll do a complete overview from the beginning of a frame to the end.

I'll do this during my spare time, so this is not gonna be completed soon.

But at least this post will not be useless as just an introduction could be.

Reply:
Alright perfect, let me know when the complete version of the article is committed then!